src/VersionUpdater/Installer.php

```diff
@@ -0,0 +1,143 @@
+<?php
+
+/*
+ * This file is part of Psy Shell.
+ *
+ * (c) 2012-2022 Justin Hileman
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Psy\VersionUpdater;
+
+use Psy\Exception\ErrorException;
+
+class Installer
+{
+    /**
+     * @var string
+     */
+    protected $installLocation;
+
+    /**
+     * @var string
+     */
+    protected $tempDirectory;
+
+    public function __construct(string $tempDirectory = null)
+    {
+        $this->tempDirectory = $tempDirectory ?: \sys_get_temp_dir();
+        $this->installLocation = \Phar::running(false);
+    }
+
+    /**
+     * Public to allow the Downloader to use the temporary directory if it's been set.
+     *
+     * @return string
+     */
+    public function getTempDirectory(): string
+    {
+        return $this->tempDirectory;
+    }
+
+    /**
+     * Verify the currently installed PsySH phar is writable so it can be replaced.
+     *
+     * @return bool
+     */
+    public function isInstallLocationWritable(): bool
+    {
+        return \is_writable($this->installLocation);
+    }
+
+    /**
+     * Verify the temporary directory is writable so downloads and backups can be saved there.
+     *
+     * @return bool
+     */
+    public function isTempDirectoryWritable(): bool
+    {
+        return \is_writable($this->tempDirectory);
+    }
+
+    /**
+     * Verifies the downloaded archive can be extracted with \PharData.
+     *
+     * @param string $sourceArchive
+     *
+     * @return bool
+     */
+    public function isValidSource(string $sourceArchive): bool
+    {
+        if (!\class_exists('\PharData')) {
+            return false;
+        }
+        $pharArchive = new \PharData($sourceArchive);
+
+        return $pharArchive->valid();
+    }
+
+    /**
+     * Extract the "psysh" phar from the archive and move it, replacing the currently installed phar.
+     *
+     * @param string $sourceArchive
+     *
+     * @return bool
+     */
+    public function install(string $sourceArchive): bool
+    {
+        $pharArchive = new \PharData($sourceArchive);
+        $pharArchive->extractTo($this->tempDirectory, ['psysh'], true);
```

Comment:
we should probably extract this to a `tempnam` directory _inside_ the temp dir, because there just might be a folder called `psysh` there already.

```
PharData::extractTo(/var/folders/lx/7kfh1rt500z6yphfzdm38f6w0000gn/T/psysh): Failed to open stream: Is a directory
```

Reply:
Thanks for catching this. I've pushed a fix that also addresses some incorrect `tempnam()` usage in the download classes!